### PR TITLE
Respect virtualenvs.in-project setting, if .venv is found in project folder

### PR DESCRIPTION
**Pull request by finswimmer**
At the moment, when there is a folder called `.venv` within the project folder, `poetry` will use it as the path to the venv regardless whether `virtualenvs.in-project` is set to `true` or not. This leads to confusions as described in #1770 and #2756.

With this PR poetry will check if  `virtualenvs.in-project` is set to `true` if it finds a `.venv` folder.

Closes: #1770
Closes: #2756

### CODE REVIEW

File: poetry/utils/env.py [hunk 1]
```diff
@@ -348,9 +347,10 @@ def get(self, reload=False):  # type: (bool) -> Env
         if not in_venv or env is not None:
             # Checking if a local virtualenv exists
             if (cwd / ".venv").exists() and (cwd / ".venv").is_dir():
-                venv = cwd / ".venv"
+                if self._poetry.config.get("virtualenvs.in-project"):
```
Comment (abn) [inline poetry/utils/env.py:350]: Should we handle `None` as "use detected environment"?

File: tests/installation/test_chef.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 from packaging.tags import Tag
+
```
Comment (abn) [inline tests/installation/test_chef.py:2]: Is this a related change? Same as the ones below.

File: poetry/utils/env.py (hunk 1 above)
Comment (finswimmer) [reply poetry/utils/env.py:350]: Didn't know that we can set it to `None` :) As @sdispater suggested in https://github.com/python-poetry/poetry/issues/1770#issuecomment-589642728 I could change it to detect `.venv` in case `in-project` is set to `true` or `None`. But than I guess I will update the docs as well, to the behavior clear.

File: tests/installation/test_chef.py (hunk 2 above)
Comment (finswimmer) [reply tests/installation/test_chef.py:2]: I didn't introduce these changes intentional. This is what `isort` has done. Don't know why.

File: poetry/utils/env.py (hunk 1 above)
Comment (abn) [reply poetry/utils/env.py]: It's the case when it was not explicitly set right? Unless we ensure that not set means true by default.

File: tests/installation/test_chef.py (hunk 2 above)
Comment (abn) [reply tests/installation/test_chef.py:2]: Weird. I guess isort ran outside of pre-commit.

Comment (finswimmer) [reply tests/installation/test_chef.py]: hm, yes. I'm used to run `isort` manually. Now I've run it via pre-commit and at least the tests passed here. But I have the strong feeling, that the pre-commit version doesn't respect the configuration in the `pyproject.toml`. Can you confirm this?

Comment (abn) [reply tests/installation/test_chef.py]: Hmm, it might be that we never ran `isort` on the entire code base after the config was changed. It might be a good idea to do a PR with `isort` run on the whole codebase.

File: tests/utils/test_env.py [hunk 3]
```diff
@@ -74,17 +74,35 @@ def test_virtualenvs_with_spaces_in_their_path_work_as_expected(tmp_dir, manager
...
+
+    venv_folder = Path(poetry.file.parent) / ".venv"
+    venv_folder.mkdir()
+    poetry.config.config["virtualenvs"]["in-project"] = False
+
+    venv = manager.get()
+
+    assert venv.path != venv_folder
+
+    shutil.rmtree(str(venv_folder))
```
Comment (abn) [inline tests/utils/test_env.py:105]:
```suggestion
    shutil.rmtree(venv_folder.as_posix())
```

File: tests/utils/test_env.py [hunk 4]
```diff
@@ -74,17 +74,35 @@ def test_virtualenvs_with_spaces_in_their_path_work_as_expected(tmp_dir, manager
...
+
+
+def test_env_get_venv_with_venv_folder_present(manager, poetry):
+    if "VIRTUAL_ENV" in os.environ:
+        del os.environ["VIRTUAL_ENV"]
+
+    venv_folder = Path(poetry.file.parent) / ".venv"
+    venv_folder.mkdir()
+    poetry.config.config["virtualenvs"]["in-project"] = False
```
Comment (abn) [inline tests/utils/test_env.py:99]: Can we not merge these two together with an additional parameter?

File: docs/docs/configuration.md [hunk 5]
```diff
@@ -112,6 +112,10 @@ Defaults to `true`.
 Create the virtualenv inside the project's root directory.
 Defaults to `false`.
 
+If set to `true`, the virtualenv wil be created and expected in a folder named `.venv` within the root directory of the project.
+
+If set to `None`, `poetry` will use the virtualenv from the `.venv` directory only if available.
```
Comment (abn) [inline docs/docs/configuration.md:117]:
```suggestion
If not set explicitly, `poetry` will use the virtualenv from the `.venv` directory only if available.
```

File: tests/utils/test_env.py (hunk 3 above)
Comment (finswimmer) [reply tests/utils/test_env.py:105]: Hm, would this work on windows?

File: tests/utils/test_env.py (hunk 4 above)
Comment (finswimmer) [reply tests/utils/test_env.py:99]: I thought about this too, but decided to keep them separated to show the different behavior more explicit.

File: docs/docs/configuration.md (hunk 5 above)
Comment (finswimmer) [reply docs/docs/configuration.md:117]: This might be confusing, because the default is `false`. One could interpret "not set explicitly" as "default not changed".

File: tests/utils/test_env.py (hunk 3 above)
Comment (abn) [reply tests/utils/test_env.py:105]: Lets leave this for now; but in theory yes it should since windows is POSIX compliant. And the use of `str()` if i recall can have some issues in various environments. Although, recently I have discovered using `pathlib2`'s `as_posix` method in `<=3.5` has other issues too. Cannot win!

File: tests/utils/test_env.py (hunk 4 above)
Comment (abn) [reply tests/utils/test_env.py:99]: Would be great if we were making the test suites easier to maintain by removing code duplication. But not a show stopper at the moment.

File: docs/docs/configuration.md (hunk 5 above)
Comment (abn) [reply docs/docs/configuration.md:117]: The ens user should never set `None`. By default it is `false` because when we attempt to get the config we get `None` because it is not set. This distinction is something we should handle internally hence the suggestion.

File: tests/utils/test_env.py (hunk 3 above)
Comment (abn) [inline tests/utils/test_env.py:105]:
```suggestion
@pytest.fixture
def in_project_venv_dir(poetry):
    os.environ.pop("VIRTUAL_ENV", None)
    venv_dir = poetry.file.parent.joinpath(".venv")
    venv_dir.mkdir()

    try:
        yield venv_dir
    finally:
        venv_dir.rmdir()


@pytest.mark.parametrize("in_project", [True, False, None])
def test_env_get_venv_with_venv_folder_present(manager, poetry, in_project_venv_dir, in_project):
    poetry.config.config["virtualenvs"]["in-project"] = in_project
    venv = manager.get()

    if in_project is False:
        assert venv.path != in_project_venv_dir
    else:
        assert venv.path == in_project_venv_dir
```

Something like this might be cleaner.

**Review by abn — CHANGES_REQUESTED** [review]
Minor suggestions. Otherwise looks good.

File: poetry/utils/env.py [hunk 6]
```diff
@@ -347,9 +347,10 @@ def get(self, reload=False):  # type: (bool) -> Env
         if not in_venv or env is not None:
             # Checking if a local virtualenv exists
             if (cwd / ".venv").exists() and (cwd / ".venv").is_dir():
-                venv = cwd / ".venv"
+                if self._poetry.config.get("virtualenvs.in-project") is not False:
```
[the remark this answers is not in the document]
Comment (finswimmer) [reply poetry/utils/env.py]: Make sense.

File: tests/utils/test_env.py (hunk 3 above)
Comment (finswimmer) [reply tests/utils/test_env.py]: I like it :+1:

File: docs/docs/configuration.md [hunk 7]
```diff
@@ -112,6 +112,10 @@ Defaults to `true`.
 Create the virtualenv inside the project's root directory.
 Defaults to `false`.
```
Comment (abn) [inline docs/docs/configuration.md:113]: Should have caught this earlier, we default to `unset` correct?

File: docs/docs/configuration.md [hunk 8]
```diff
@@ -112,6 +112,10 @@ Defaults to `true`.
...
 
+If set to `true`, the virtualenv wil be created and expected in a folder named `.venv` within the root directory of the project.
+
+If not set explicitly, `poetry` will use the virtualenv from the `.venv` directory only if available.
```
Comment (abn) [inline docs/docs/configuration.md:117]:
```suggestion
If not set explicitly, `poetry` will use the virtualenv from the `.venv` directory when one is available. If set to `false`, `poetry` will ignore any existing `.venv` directory.
```

**Review by abn — COMMENTED** [review]
Missed the documentation in the last pass.

File: docs/docs/configuration.md (hunk 7 above)
Comment (finswimmer) [reply docs/docs/configuration.md:113]: No, the default is still `false`. (At least I haven't changed it). And in my opinion we shouldn't change this behavior.

Comment (abn) [reply docs/docs/configuration.md:113]: Well the issue is that with this change we break existing behaviour. Currently if default (`false`), and a `.venv` directory is found `poetry` uses it. This behaviour can be replicated by configuring default to to be `None`  with the new changes. And hence why I reckon we should update the documentation and default value to reflect this. Otherwise, we end up creating new venvs where this "misfeature" was relied upon.

Comment (finswimmer) [reply docs/docs/configuration.md:113]: While it is correct that this PR changes the current behavior and it could be replicate by setting to `None`, I don't think we should change the default value. Before this PR the behavior was surprising to many people and now it behave like the most people would expect. So it's more like a fix and not like a new feature.

But it's definitely something that should be stated out clearly in the release notes.

Comment (abn) [reply docs/docs/configuration.md:113]: The problem with that is that the current users who a) have not explicitly ever set it to `true` or `false`, and b) relies on existing `.venv` directories in their projects will suddently end up creating new virtual environments after upgrade. I do not think this is acceptable. Changing the default however, ensures that this case is covered for. And there is no difference for the other scenarios.

**Review by abn — APPROVED** [review]
